File: sdk/ruby/lib/hitechcloud/resources/cloud_gpu.rb
module HiTechCloud
  # CloudGpu resource
  class CloudGpuResource < BaseResource

    # Reboot VM
    def reboot(id, data)
      @http.put("/api/service/%{id}/vms/reboot" % { id: id }, data)
    end

    # Stop VM
    def stop(id, data)
      @http.put("/api/service/%{id}/vms/stop" % { id: id }, data)
    end

    # Start VM
    def start(id, data)
      @http.put("/api/service/%{id}/vms/start" % { id: id }, data)
    end

    # Get firewall rules
    def firewall(id)
      @http.get("/api/service/%{id}/vms/firewall" % { id: id })
    end

    # Create firewall rule
    def create_firewall(id, data)
      @http.post("/api/service/%{id}/vms/firewall" % { id: id }, data)
    end

    # Delete firewall rule
    def delete_firewall(id, position)
      @http.delete("/api/service/%{id}/vms/firewall/%{position}" % { id: id, position: position })
    end

  end
end

File: sdk/python/hitechcloud/resources/cloud_gpu.py
"""Cloud GPU resource (6 endpoints)"""

from .base import BaseResource


class CloudGpuResource(BaseResource):
    """Cloud GPU: GPU instance management"""

    def reboot(self, service_id: int) -> dict:
        """PUT /api/service/{id}/vms/reboot - Reboot virtual server"""
        return self._http.put(f"/api/service/{service_id}/vms/reboot")

    def stop(self, service_id: int) -> dict:
        """PUT /api/service/{id}/vms/stop - Stop virtual server"""
        return self._http.put(f"/api/service/{service_id}/vms/stop")

    def start(self, service_id: int) -> dict:
        """PUT /api/service/{id}/vms/start - Start virtual server"""
        return self._http.put(f"/api/service/{service_id}/vms/start")

    def get_firewall_rules(self, service_id: int) -> dict:
        """GET /api/service/{id}/vms/firewall - Get firewall rules"""
        return self._http.get(f"/api/service/{service_id}/vms/firewall")

    def add_firewall_rules(self, service_id: int, rules: list) -> dict:
        """POST /api/service/{id}/vms/firewall - Add firewall rules"""
        return self._http.post(f"/api/service/{service_id}/vms/firewall", {"rules": rules})

    def remove_firewall_rule(self, service_id: int, position: int) -> dict:
        """DELETE /api/service/{id}/vms/firewall/{position} - Remove firewall rule"""
        return self._http.delete(f"/api/service/{service_id}/vms/firewall/{position}")
